### ISSUE TITLE
Streaming from assets

### ISSUE THREAD

**Issue by ryanheise** [flutter/flutter#73322]
## Use case

I would like to efficiently stream large audio assets directly from an asset bundle. To do this, I need both a stream-based access API and the ability to seek to given byte offsets. The current AssetBundle API is limited to reading the entire asset into memory. The only workaround currently seems to be to dump that memory into a file and then use file-based APIs to read/seek, but it is still unfortunate to have to both read the entire asset into memory first, and then also to go through the additional step of writing a copy of it back to storage.

## Proposal

It would be useful to have an API similar to Android's [AssetManager](https://developer.android.com/reference/android/content/res/AssetManager) which provides access to the asset via a stream with a method for seeking, subject to different access modes.

**Comment by iapicca** [comment]
@ryanheise 
according to this [stackoverflow answer](https://stackoverflow.com/a/51307450/9836706) this seems to be doable already

**Comment by ryanheise** [comment]
Every answer on that StackOverflow page demonstrates how to read the entire asset into memory which is the precisely the problem that my proposal aims to rectify.

**Comment by iapicca** [comment]
@ryanheise 
a possible solution is to use [AssetManager](https://developer.android.com/reference/android/content/res/AssetManager) passing the data to flutter via  [platform channel](https://flutter.dev/docs/development/platform-integration/platform-channels) (possibly using [pigeon](https://pub.dev/packages/pigeon))
as this approach  being platform specific  and requires to handle the assets in a very different way
imho make more sense as a separate package than as feature for flutter

**Comment by ryanheise** [comment]
To provide further context, this is for the [just_audio](https://pub.dev/packages/just_audio) plugin, an audio player that can play assets from various audio sources including URLs, files *and* assets.

By "assets" I do mean Flutter assets, and this is what most Flutter developers would take it to mean, since there are many plugins that deal with assets. It would not be a good idea for a plugin to over-complicate things for its users by having regular Flutter assets not work and to require them to learn how to use platform-specific assets. Rather, I think we would all want to *improve* Flutter itself so that it would be possible to read an asset as a stream, to avoid having to read the entire asset into memory in one shot.

Thus the proposal.

**Comment by pyrossh** [comment]
Yep this feature would be good as it would improve performance. Right now I need to load a large gzipped csv file from the assets completely into memory and then do the utf8 decoding. It would be simpler if I could just stream the gzip and utf8 decoder.

@ryanheise btw I'm using your just_audio plugin and hit the same issue as I have a giant 1GB audio file where I seek and play certain ranges within it. I would have to go with the File based approach for this for now or else split it all into 100s of small mp3s so that the memory doesn't get overloaded.
